### ISSUE TITLE
CAD/Project File Reviews

### ISSUE THREAD

**Issue by sharonyang16** [Northeastern-Electric-Racing/FinishLine#2667]
## Overview

The project file reviews feature serve to track part and assembly drawing reviews within FinishLine. Users are able to create parts tied to a project and are able to upload submissions and reviews for those parts.

## Stakeholders

Product Stakeholder: @sharonyang16 
Software Stakeholder: Peyton McKee, Sean Walker
Reference Users: Ethan Herrell, Sofia Varner

## User Story

1. As a head/lead, I want to be able to create and assign parts to other users, so that the drawing review process is tracked within FinishLine.
2. As an admin, I want to have the ability to add/edit submission guides, so that users are able to easily access up to date information on uploading submissions for a particular project and reviewers are able to easily point out common mistakes made by assignees without having to type out the same explanation every time. 
3. As a assignee, I want to be able to easily upload submissions for a part I'm creating drawings for, view any review comments left for me, and re-ping reviewers once my submission is ready to look at again, so that I can efficiently complete designs for important parts.
4. As a reviewer, I want to be able to access and review designs submitted, so that part assignees are able to complete their designs and we, as a team, are able to move forward in the manufacturing process.

## Success Metrics

Users are able to track the entire review process for a part within FinishLine.

## Rollout Plan

The product side of this feature should be completed during Summer 2024 for developers to begin implementation during late Fall 2024 or beginning of Spring 2025. This feature should be rolled out all at once.

## Out of Scope

Creating drawing markups on files/having a pen tool for reviews is out of scope for this epic, but this may be a feature added in the future.

## Background / Context

Engineers currently conduct drawing reviews for parts by creating new Confluence pages and as an effort to continue to track more of the team's work within FinishLine, this feature was suggested. As a result, we are creating a review system (with similarities to GitHub Pull requests) that allows for this review process to be conducted in FinishLine.

### Overview/Files Tab

A new "Parts Review" tab should be added to each project's page.
![Files Tab](https://github.com/user-attachments/assets/5c833de7-ac8a-4773-a5ab-1d7e51704cc8)
- Each Part Card should display the first file of the most recent submission or review, the part's name, which stage of review the part is in, the assignees and reviewers for the part, and a button to download a formatted PDF of all the iterations and comments for that part that we can use for competition
- The search filters should include assignee name, reviewer name, date uploaded, review status (i.e. Ready for Review, In Review, Reviewed, and Approved), and part tags.
- The submission guide should include a section for a Sample Drawing, FAQs, a link to an external Confluence guide and Common Mistakes. These will be configurable for Admins (see Admin Tools section). The guide should toggled off by default for admins, heads, and leads and toggled on for other roles. The toggle value should be persistent (can be done using local storage).
![Files Tab - Guide Open](https://github.com/user-attachments/assets/b7ace529-11be-4643-b134-9151ee00abb5)
- When clicking the "New" button, users should be given the options to upload a new submission, upload a new file based review (will be further explained below), and to create a new part. Where the drop down appearing similar to this: 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; ![Group 430](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/e44b83a6-0055-413c-8f8f-5b4ce153c8cd) 
The "New Submission" form is used for when an assignee of a part needs to upload a new files. The form should allow users to select the part they're creating a submission for (with the drop down only showing unapproved parts where the current user is an assignee), to upload file(s), to give their submission a name, and leave any additional comments for reviewers (optional).
![New Submission Popup](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/22eeb0ca-5d31-4cec-8dd7-c16fc67d8eaf)
The "New Review" form is used for when a reviewer needs to upload a file or files as part of their review for a submission. The form should allow the user to select the part (with the drop down only showing unapproved parts where the current user is a reviewer), to upload file(s), to select the submission that the review is targeting, and leave any additional comments as part of their review.
![New Review popup](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/ed82a50c-a2fc-49f9-9171-50739a96e97a)
The "New Part" form is used to create a new part to be tracked and reviewed. Parts should follow the naming convention as the [NER Part Numbers Google Sheet](https://docs.google.com/spreadsheets/d/1_WRncGvwAv0z4CI_1dNmEjzXzphcYhTtymyae6SjH8w/edit#gid=0) where the names are in the format of Project_CommonName_Index-CarNumber. For example: <img width="700" alt="image" src="https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/108b78e3-e585-4b2a-bb9b-e96de25a1d36">
The New Part form should have the user manually enter the index and common name for the part, but the feature should be able to infer the project and car number to generate the full name for the part. The form should allow the user to assign assignees and reviewers for the part along with setting an optional description and tags associated with the part.
![New Part Popup](https://github.com/user-attachments/assets/32047b44-55b5-496e-a1d3-6c034aa6a170)

A "My Parts Under Review" table should appear if the current user is assigned to a part that hasn't been approved by a reviewer (i.e. any part assigned that's does not have the "Approved" status), a "Parts for me to Review" table should appear for the current user if they are set as a reviewer on a part and has not reviewed the most recent submission, and a "All Parts Under Review" table should appear for heads/leads of a project as they are the user group that's primarily reviewers. All these tables should be in one row.

Here's what the tables would look like if all 3 were to appear for the user:
![Files Tab - Leadership](https://github.com/user-attachments/assets/761d3fc9-03df-4aca-8eb7-850b2288db55)

Here's what they would look like if only 2 of the 3 were present:
![Files Tab - Leadership (No _My Parts_)](https://github.com/user-attachments/assets/87b25e79-feee-4e99-99c9-6c6ae5fbc28d)

And here's what they would look like if only 1 of the tables are present (this is what most members would see if they have an assigned part that's not approved yet):
![Files Tab - 1 table](https://github.com/user-attachments/assets/ec8f1e58-ae3d-4118-b2bd-a8b13d4a5c69)

To wrap this section up, this is an overview/scroll view of all the mentioned sections on the Parts Review tab:
![Files Tab - Guide Open](https://github.com/user-attachments/assets/40f6b654-c85d-4cca-9a3d-1bd96534071c)

### Part Details 
After creating a new part, the details page for that part should be empty and appear like this: 
![File Details](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/5d89586f-cdbd-4aab-b5f1-9ffa3087f992)

The optional description and tags for a part should appear first in the Overview section and before the review status label and chip
![File Details - Ready for Review optional](https://github.com/user-attachments/assets/f6ff7fd4-7c05-4b6a-8e78-8c916dbec13a)

Once the first submission is uploaded, the submission should be viewable on the left and the part should be automatically be set as ready for review.
![File Details - Ready for Review 1](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/5c852334-3aad-46f9-8f20-55ca013af765)

Once a reviewer begins a review (i.e. started a review, but did not finish/hit "Save as Draft"), the part should be set as in review.
![File Details - In Review](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/6f57cc1e-ed7e-4082-8a93-b620cb290828)

Once a review is submitted for the most recent submission from an assignee, the part should be set as reviewed. 

If the reviewer decided to review the submission using only in app markups, the file should now have points on where the reviewer decided leave comments. When hovering over the points, there should be a popup of what comment was left there, including which reviewer left that comment.
![File Details - Ready for Review (Submitter View)](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/4b36876f-4e8f-4efc-9475-4b906d21a68a)

If the reviewer decided to review the submission by downloading the submission, creating markups using an external resource, and uploading the new marked up version of the file, a new "page" should be created in the part and be named "[Name of Reviewed Submission] Review". In this mockup, a user should be able to access the review from the "Submission #&#8203;1" page by clicking the "Next" button (shown in the first image) to access the new "page" created by the review (shown in the second image). Note that the middle section (titled "Details for [Page]") changes depending on the "page" the user is currently viewing.
![File Details - Reviewed File Page 1](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/d8dd1dc9-bfd9-4682-8f39-0c3dbc80df60)
![File Details - Reviewed Page 2](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/03844270-e941-4e61-a5fa-7699f7af0eda)

Once a new submission is uploaded, the part should be set to review to review again and the assignee can rerequest a review from a reviewer. There should be a "Next" and "Previous" button available when relevant and there was something submitted (whether a submission or review) before or after the current "page". 
![File Details - Cycle](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/e54b5eca-471b-4bc8-afc2-34650466ac99)

Reviewers should be able to approve a submission using the actions menu and the status of the part should be set as approved. The Action button should be changed to allow leadership to reopen for any reason.
![File Details - Approved](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/8aef2d94-6956-4d00-be89-0ac4a6655913)

These should be the options when clicking the actions button in the top right corner. 
![details dropdown](https://github.com/user-attachments/assets/476a4207-00da-4e70-8fbb-237f99f4807d)

- Assignees should be restricted to being only able to access the Upload Submission option, which would bring up the "New Submission" modal from before, and the Edit Submission option only if the submission they are trying to edit has not been reviewed yet
- Reviewers should be able to access the Review and Approve options, with the review option bringing them to the review page (detailed further down) and the approve option bringing up this modal:
![New Review popup](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/28a69c17-25b0-4b34-a117-a850f2416dd6)

- Heads/leads should be able to access the Edit and Delete options, with the edit option bringing up the "New Part" Modal from before (but with the title being Edit Part) and allow the user to change the part name and assignees and reviewers and the delete option would show a confirmation message to ensure a part is not being deleted by mistake. 

### Reviewing

When selecting the review option on a part, the reviewer should be met with this screen. The reviewer should select the submission that they are targeting in this review to begin the review process. The question mark icon next to "Review Status" should show a popup telling the reviewer that they need to upload at least one file or mark up the selected submission with at least one comment (either a custom comment or a common mistake comment) in order to finish/submit the review. 
![Review File](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/101f66e9-e7c7-46bf-8f12-8f8ebfca19d1)

When the custom comment/text tool is selected, the cursor should appear as a crosshair when hovering over the file area. Once the user clicks, a popup should appear from the spot selected which has the user put in a title/shorter description and a longer description for what they want to comment on.
![Review File - Text](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/e8baa972-21e7-42a7-8cc4-a3422c8ec7ee)

Once the reviewer hits done, the spot should be clickable and the popup with the comment would reappear and be deletable and editable.  Additionally, after a comment is submitted, the "In-App Markups" section under "Review Status" should update to let the user know which pages already have comments and how many. 
![Review File - Text Hover](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/9f1a6b1e-906a-472e-8368-7bc7c98a76c4)

The common mistake tool should work similarly to the text tool, but instead, the popup form would have the user select from a list of common mistakes for the project.
![Review File - Common mistake](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/7b7d020f-ebd8-4b65-b94f-ca1ada3eaa28)

In this dropdown, reviewers should have the option to add a common mistake if the one they're looking for doesn't exist
![Group 454](https://github.com/user-attachments/assets/0e7f6cb4-9dd4-469f-95c7-1709bb077c53)

The popup after the user hits done would also be similar to the text tool's.
![Review File - Common mistake Hover](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/c8a985f0-a54c-4aa0-a941-dfffe167884d)

Clicking the last tool in the toolbar should allow users to upload files for the review. Once uploaded, the "File(s) Uploaded" section under "Review Status" should update with the files uploaded and users should be able to delete individual files if uploaded by accident.
![Review File - File upload](https://github.com/Northeastern-Electric-Racing/FinishLine/assets/92332953/4f769537-c5ff-49b0-9993-2640eb9579b5)

### Admin Tools
There should be a new Parts Review Config section added to the Projects tab of the Admin Tools
![Admin Tools Project Selected](https://github.com/user-attachments/assets/614e9ed5-bae7-42eb-a591-5c8ecf8bd999)
- The Project Name Abbreviations section should allow admins to select a project and add an abbreviation for the project name to be used to populate part names for parts under that project
- The Part Tags section should allow admins to manage tags to be attached to parts. Tags should not be able to be deleted if any existing part has that tag.
- The Confluence Drawing Guide section should allow admins to change the guide link in the Submission Guide in case the URL changes in the future 
- The Sample Drawing section should allow admins to change the sample drawing displayed in the Submission Guide and should be able to be multiple files
- The FAQ section should allow admins to manage the FAQ section of the Submission Guide
- The Common Mistakes section should allow admins to manage the common mistakes used in reviews. Admins should star the common mistakes that they want to be displayed in the Submission Guide and the tooltip icon next to this section's header should make note of this

[Here](https://docs.google.com/document/d/1fyK9omGt4lw2WrN1rUN8wEhd3Kh03-fphgzsxnRqyOk/edit) is a list of initial part tags and common mistakes we can add

## Tickets

 - [x] #3125
 - [x] #3160 
 - [x] #3165 
 - [x] #3161
 - [x] #3163 
 - [x] #3169 
 - [x] #3170
 - [x] #3164
 - [x] #3166 
 - [x] #3157
 - [x] #3186
 - [x] #3187
 - [x] #3188
 - [x] #3199 
 - [x] #3236 
 - [x] #3265 
 - [x] #3238 
 - [x] #3240 
 - [x] #3241 
 - [x] #3260 
 - [x] #3274
 - [x] #3275
 - [x] #3276
 - [x] #3202 
 - [x] #3203 
 - [x] #3204 
 - [x] #3282
 - [x] #3284 
- [x] #3287
- [x] #3289 
- [x] #3290
- [x] #3311
- [x] #3316 
- [x] #3335
- [x] #3337
- [x] #3338
- [x] #3339 
- [x] #3340
- [x] #3351 
- [x] #3368 
- [x] #3369 
- [x] #3370 
- [x] #3384
- [x] #3385 
- [x] #3389
- [x] #3392 
- [x] #3393
- [x] #3394 
- [x] #3408 
- [x] #3409
- [x] #3426 
- [x] #3434 
- [x] #3435 
- [x] #3458
